core/src/main/java/com/scalar/db/storage/dynamo/DynamoTableMetadataManager.java

```diff
@@ -127,19 +153,174 @@ private DataType convertDataType(String columnType) {
+      throw new StorageRuntimeException(
+          "adding meta data for table "
+              + getFullTableName(namespacePrefix, namespace, table)
+              + " failed",
+          e);
+    }
   }
 
   @Override
   public Set<String> getTableNames(String namespace) {
-    // TODO To implement
-    throw new UnsupportedOperationException();
+    Set<String> tableSet = new HashSet<>();
+    try {
+      ListTablesResponse listTablesResponse = client.listTables();
+      List<String> tableNames = listTablesResponse.tableNames();
+      for (String tableName : tableNames) {
+        if (tableName.startsWith(getFullNamespaceName(namespacePrefix, namespace))) {
+          tableSet.add(tableName);
+        }
+      }
+    } catch (DynamoDbException e) {
+      throw new StorageRuntimeException("retrieving the table names failed", e);
+    }
+    return tableSet;
+  }
+
+  private void createMetadataTableIfNotExist() throws StorageRuntimeException {
+    if (!metadataTableExists()) {
+      CreateTableRequest.Builder requestBuilder = CreateTableRequest.builder();
+      List<AttributeDefinition> columnsToAttributeDefinitions = new ArrayList<>();
+      columnsToAttributeDefinitions.add(
+          AttributeDefinition.builder()
+              .attributeName(TABLE)
+              .attributeType(ScalarAttributeType.S)
+              .build());
+      requestBuilder.attributeDefinitions(columnsToAttributeDefinitions);
+      requestBuilder.keySchema(
+          KeySchemaElement.builder().attributeName(TABLE).keyType(KeyType.HASH).build());
+      requestBuilder.provisionedThroughput(
+          ProvisionedThroughput.builder()
+              .readCapacityUnits(METADATA_TABLE_RU)
+              .writeCapacityUnits(METADATA_TABLE_RU)
+              .build());
+      requestBuilder.tableName(
+          getFullTableName(namespacePrefix, METADATA_NAMESPACE, METADATA_TABLE));
+
+      try {
+        client.createTable(requestBuilder.build());
+      } catch (DynamoDbException e) {
+        throw new StorageRuntimeException("creating meta data table failed");
+      }
+
+      while (true) {
+        try {
+          Thread.sleep(CREATING_WAITING_TIME);
+          DescribeTableRequest describeTableRequest =
+              DescribeTableRequest.builder()
+                  .tableName(getFullTableName(namespacePrefix, METADATA_NAMESPACE, METADATA_TABLE))
+                  .build();
+          DescribeTableResponse describeTableResponse = client.describeTable(describeTableRequest);
+          if (describeTableResponse.table().tableStatus() == TableStatus.ACTIVE) {
+            break;
+          }
+        } catch (DynamoDbException | InterruptedException e) {
+          throw new StorageRuntimeException("getting table description failed", e);
+        }
+      }
+    }
+  }
+
+  private boolean metadataTableExists() throws StorageRuntimeException {
+    boolean tableExist = false;
+    try {
+      ListTablesResponse listTablesResponse = client.listTables();
```

Comment:
Maybe it's better to use `client.describeTable()` and handle `ResourceNotFoundException` when the table is not found?

Reply:
Thanks for the suggestion. Apply in [83b7020](https://github.com/scalar-labs/scalardb/pull/301/commits/83b7020c4783675821630ff67827a70b29f5692d). Other suggestions were applied as well.